reference/src/client/src/sa_provider_signature.c

```diff
@@ -0,0 +1,1202 @@
+                    algorithm_id_length = sizeof(rsa_sha1_oid);
+                    break;
+
+                case NID_sha256:
+                    algorithm_id = rsa_sha256_oid;
+                    algorithm_id_length = sizeof(rsa_sha256_oid);
+                    break;
+
+                case NID_sha384:
+                    algorithm_id = rsa_sha384_oid;
+                    algorithm_id_length = sizeof(rsa_sha384_oid);
+                    break;
+
+                case NID_sha512:
+                    algorithm_id = rsa_sha512_oid;
+                    algorithm_id_length = sizeof(rsa_sha512_oid);
+                    break;
+
+                default:
+                    break;
+            }
+        } else if (signature_context->key_data->type == EVP_PKEY_EC) {
+            switch (EVP_MD_get_type(signature_context->evp_md)) {
+                case NID_sha1:
+                    algorithm_id = ecdsa_sha1_oid;
+                    algorithm_id_length = sizeof(ecdsa_sha1_oid);
+                    break;
+
+                case NID_sha256:
+                    algorithm_id = ecdsa_sha256_oid;
+                    algorithm_id_length = sizeof(ecdsa_sha256_oid);
+                    break;
+
+                case NID_sha384:
+                    algorithm_id = ecdsa_sha384_oid;
+                    algorithm_id_length = sizeof(ecdsa_sha384_oid);
+                    break;
+
+                case NID_sha512:
+                    algorithm_id = ecdsa_sha512_oid;
+                    algorithm_id_length = sizeof(ecdsa_sha512_oid);
+                    break;
+
+                default:
+                    break;
+            }
+        } else if (signature_context->key_data->type == EVP_PKEY_ED25519) {
+            algorithm_id = eddsa_ed25519_oid;
+            algorithm_id_length = sizeof(eddsa_ed25519_oid);
+        } else if (signature_context->key_data->type == EVP_PKEY_ED448) {
+            algorithm_id = eddsa_ed448_oid;
+            algorithm_id_length = sizeof(eddsa_ed448_oid);
+        }
+
+        if (!OSSL_PARAM_set_octet_string(param, algorithm_id, algorithm_id_length)) {
+            ERROR("OSSL_PARAM_set_octet_string failed");
+            return 0;
+        }
+    }
+
+    param = OSSL_PARAM_locate(params, OSSL_SIGNATURE_PARAM_PAD_MODE);
+    if (param != NULL) {
+        if (param->data_type == OSSL_PARAM_INTEGER) {
+            if (!OSSL_PARAM_set_int(param, signature_context->padding_mode)) {
+                ERROR("OSSL_PARAM_set_int failed");
+                return 0;
+            }
+        } else if (param->data_type == OSSL_PARAM_UTF8_STRING) {
+            if (signature_context->padding_mode == RSA_PKCS1_PADDING) {
+                if (OSSL_PARAM_set_utf8_string(param, OSSL_PKEY_RSA_PAD_MODE_PKCSV15) != 1) {
+                    ERROR("OSSL_PARAM_set_utf8_string failed");
+                    return 0;
+                }
+            } else if (signature_context->padding_mode == RSA_PKCS1_PADDING) {
```

Comment:
Duplicate comparison. I think you're looking to check if it is PSS mode padding.

Reply:
Done